freqtrade/strategy/strategy_helper.py

```diff
@@ -81,30 +80,26 @@ def stoploss_from_open(
     The requested stop can be positive for a stop above the open price, or negative for
     a stop below the open price. The return value is always >= 0.
 
-    Returns 0 if the resulting stop price would be above the current price.
+    Returns 0 if the resulting stop price would be above/below (longs/shorts) the current price
 
     :param open_relative_stop: Desired stop loss percentage relative to open price
     :param current_profit: The current profit percentage
+    :param for_short: When true, perform the calculation for short instead of long
     :return: Stop loss value relative to current price
     """
 
-    # formula is undefined for current_profit -1, return maximum value
-    if current_profit == -1:
+    # formula is undefined for current_profit -1 (longs) or 1 (shorts), return maximum value
+    if (current_profit == -1 and not for_short) or (for_short and current_profit == 1):
         return 1
 
     if for_short is True:
-        # TODO-lev: How would this be calculated for short
-        raise OperationalException(
-            "Freqtrade hasn't figured out how to calculated stoploss on shorts")
-        # stoploss = 1-((1+open_relative_stop)/(1+current_profit))
+        stoploss = -1+((1-open_relative_stop)/(1-current_profit))
```

Comment:
Can you give some explanation for why this is the formula? Or provide a source?

Reply:
Sure, I deduced it with the following:

Defining the following:
`current_price = open_price * (1 - current_profit)` (`current_profit` positive, current price lower)
`open_stoploss_price = open_price * (1 - open_stoploss_ratio)` (`open_stoploss_ratio` negative, current price higher)

`current_relative_stoploss = 1-(open_stoploss_price/current_price)`
Given that `current_relative_stoploss` is the ratio of the current price and the stoploss price when the trade was open. So, replacing `current_price` and `open_stoploss_price` with they're definitions:
`
current_relative_stoploss = 1 - ((open_price * (1 - open_stoploss_ratio))/(open_price * (1 - current_profit)))
`
Factoring out `open_price` gives:
`current_relative_stoploss = 1 - ((1 - open_stoploss_ratio) / (1 - current_profit))`

And finally, I change sign of the whole formula:
`current_relative_stoploss = -1 + ((1 - open_stoploss_ratio) / (1 - current_profit))`

This last one may be the most controversial part, but it's needed to respect the previous sign convention established by this function: for longs, `open_stoploss_ratio` is negative when `open_price` is above the `open_stoploss_price`, but this function returns a positive value when `current_price` is above `open_stoploss_price`. So, for shorts, the opposite criteria is taken, negative when the `open_price` is above the `open_stoploss_price`, and returns positive value value when `current_price` is below `open_stoploss_price`.

So, classical example using the short position tool from TradingView:
![image](https://user-images.githubusercontent.com/15948561/149630332-bcbe32ec-79e7-4b8b-9501-f9f1a4a4add7.png)
`open_price` = 10000 (grey)
`open_stoploss_price` = 11000 (red) -> `open_stoploss_ratio` = 0.1
`current_price` = 8000 (green) -> `current_profit` = 0.2
Now, in this same tool, calculate stoploss from `current_price` to `open_stoploss_price`:
![image](https://user-images.githubusercontent.com/15948561/149630500-0a23de50-98a7-43f3-bab9-eb34eec1ac76.png)

And if I perform this calculation in the formula yields the same value:
`current_relative_stoploss = -1 + ((1 - (-0.1)) / (1 - 0.2)) = 0.375`

Also, the test pass because the same definitions is made in it.

Reply:
I think you'll need to be careful with the docstrings here. It may be missleading, and should really be compared with the one in `develop`.

> This last one may be the most controversial part, but it's needed to respect the previous sign convention established by this function

i don't really understand this.
You're talking about prices throughout the above post ... but absolute prices are not relevant for this function (They are for `stoploss_from_absolute()` - but not for `stoploss_from_open()` (which we're talking about here).

`current_profit` will always have the same sign - `5`% when you're green - `-5`% when you're red - if it's a long or a short won't matter.
`open_relative_stop` is `Desired stop loss percentage relative to open price`.

so the purpose of this function is to simplify implementation of custom stoploss levels.

calling `stoploss_from_open(0.02, current_profit)` will return the correct response to have stoploss 2% above the open rate (this needs to be calculated relative to current profit, hence the helper function).

if you call `stoploss_from_open(0.0, current_profit)` - that's a "break even" function - which will place the stoploss at the open rate (the return value of `custom_stoploss` is the difference to current price).

Reply:
Yes, so I was a bit loss in the usage of this function, but after further review, I think the implement is correct, even though maybe docstring could be improved. And also, there's where my confusion with the sign convention comes from.

When I talk about open_stoploss_price, it's just a ratio with the open_price, so this fulfils the purpose you described.

And also, I speak about prices because the definitions are needed to calculate the formula, but of course, they're factored out and they're not used at all, they just serves as example to make sure I did the math correctly.